File: Swift-C++-Testsuite/test/error-handling/Swift/Defer.swift
@main
struct MyMain {

    static func main() {
        try? processFile(filename: "testfile.txt")
    }
    static func processFile(filename: String) throws {
        if exists(filename) {
            let file = open(filename)
            defer {
                file.close()

            }
            let line = try file.readline()
            // Work with the file.

            // close(file) is called here, at the end of the scope.
        }
    }
    static func exists(_ fName: String) -> Bool {
        return true
    }
    static func open(_ fName: String) -> File {
        return File(name: fName)
    }

}

struct File {
    let name: String

    func readline() throws -> String? {
        return "test"
    }

    func close() -> String {
        return "closed"
    }
}

File: Swift-C++-Testsuite/test/error-handling/C++/Defer.cpp

#include <iostream>
#include <string>

struct File {
    std::string name;
 
    std::string readline(){
        return "test";
    }

    ~File() {
        close();
    }
private: 
     std::string close() {
        return "closed";
    }
};
 
    static bool exists(std::string fName) {
        return true;
    }
 
    static File open(std::string fName) {
        return File{fName};
    }
 
    static void processFile(std::string filename) {
        if (exists(filename)) {
            File file = open(filename);


                std::string line = file.readline();
                // work with the file
      
        }
    }



int main(){
           processFile("testfile.txt");
}